### PR TITLE
Quda work merge master

### PR DESCRIPTION
**Pull request by kostrzewa**
Here's my attempt to synchronise master and quda_work. I've not tested compilation and hence also not done any regression tests. To be honest, I never wanted to touch this code again...

### CODE REVIEW

**Comment by kostrzewa** [comment]
It would be nice if someone could set up github actions to at least automatically test the build system...

**Comment by sunpho84** [comment]
what do you mean? isn't travis cehcking the build?

**Comment by kostrzewa** [comment]
> what do you mean? isn't travis cehcking the build?

that has been broken for ages.

File: tmlqcd_config.h [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+ // these PACKAGE* defines
+ #undef PACKAGE
+ #undef PACKAGE_BUGREPORT
+ #undef PACKAGE_NAME
+ #undef PACKAGE_STRING
+ #undef PACKAGE_TARNAME
+ #undef PACKAGE_VERSION
+
+ // now we include the autoconf-generated header which hopefully correctly sets all the relevant macros
+ #include "tmlqcd_config_internal.h"
```
Comment (sunpho84) [inline tmlqcd_config.h:22]: This should be `include/tmlqcd_config_internal.h`

**Comment by kostrzewa** [comment]
Anyway, this builds now on my machine but one has to be careful to create a new build directory (or appropriately empty the current one) because there is a change to the files generated in `${top_build_dir}/include`.

File: tmlqcd_config.h (hunk 1 above)
Comment (kostrzewa) [reply tmlqcd_config.h:22]: doesn't really matter, `${top_build_dir}/include` is passed as an include directory in `configure.in`

Comment (kostrzewa) [reply tmlqcd_config.h:22]: god I hate this code...

File: phmc.c [hunk 2]
```diff
@@ -18,13 +18,14 @@
  ***********************************************************************/
 
 #ifdef HAVE_CONFIG_H
-# include<tmlqcd_config.h>
+# include<config.h>
```
Comment (sunpho84) [inline phmc.c]: Why is this pointing to config.h?

Comment (kostrzewa) [reply phmc.c]: merge error, corrected

File: tmlqcd_config.h (hunk 1 above)
Comment (sunpho84) [reply tmlqcd_config.h:22]: if I create a build directory, and then compile out of place, it does not find the tmlqcd_config_internal.h

Comment (kostrzewa) [reply tmlqcd_config.h:22]: works for me, I always compile outside the build directory. Do you have the latest commit here?

Comment (kostrzewa) [reply tmlqcd_config.h:22]: I mean outside the source directory.

**Comment by kostrzewa** [comment]
Doing the following:

```
CC=mpicc CXX=mpicxx F77=f77 \                                                                                                                                
CFLAGS="-mtune=znver2 -march=znver2 -O3 -mavx2 -mfma -fopenmp" \
LDFLAGS="-fopenmp" \
~/code/tmLQCD.merge_quda_work/configure \
  --enable-mpi \
  --with-mpidimension=4 \
  --disable-sse2 --disable-sse3 \
  --with-cudadir=/usr/local/cuda/lib64 \
  --with-qudadir=$(pwd)/../quda_develop/install_dir \
  --with-limedir=$(pwd)/../lime/install_dir \
  --with-lapack="-lblas -llapack"
```

in an empty build directory works for me, but maybe there is still some compiler dependence...

File: tmlqcd_config.h (hunk 1 above)
Comment (sunpho84) [reply tmlqcd_config.h:22]: > works for me, I always compile outside the build directory. Do you have the latest commit here?

yes

**Comment by sunpho84** [comment]
> Doing the following:
> 
> ```
> CC=mpicc CXX=mpicxx F77=f77 \                                                                                                                                
> CFLAGS="-mtune=znver2 -march=znver2 -O3 -mavx2 -mfma -fopenmp" \
> LDFLAGS="-fopenmp" \
> ~/code/tmLQCD.merge_quda_work/configure \
>   --enable-mpi \
>   --with-mpidimension=4 \
>   --disable-sse2 --disable-sse3 \
>   --with-cudadir=/usr/local/cuda/lib64 \
>   --with-qudadir=$(pwd)/../quda_develop/install_dir \
>   --with-limedir=$(pwd)/../lime/install_dir \
>   --with-lapack="-lblas -llapack"
> ```
> 
> in an empty build directory works for me, but maybe there is still some compiler dependence...

yes actually it almost does, now I get 

```
../../io/utils_write_first_message.c: In function ‘write_first_messages’:
../../io/utils_write_first_message.c:27:115: error: ‘TMLQCD_PACKAGE_VERSION’ undeclared (first use in this function); did you mean ‘LIME_PACKAGE_VERSION’?
   27 |   snprintf(message, 1024, "This is the %s code for twisted mass Wilson QCD\n\nVersion %s, commit %s\n",executable,TMLQCD_PACKAGE_VERSION,git_hash);
      |                                                                                                                   ^~~~~~~~~~~~~~~~~~~~~~
      |                                                                                                                   LIME_PACKAGE_VERSION
../../io/utils_write_first_message.c:27:115: note: each undeclared identifier is reported only once for each function it appears in

```

**Comment by kostrzewa** [comment]
> yes actually it almost does, now I get

delete the file `include/tmlqcd_config.h` in your build dir

**Comment by sunpho84** [comment]
ok I see, a make distclean did the trick

File: tmlqcd_config.h (hunk 1 above)
Comment (sunpho84) [reply tmlqcd_config.h:22]: ok this is fixed

**Review by sunpho84 — APPROVED** [review]
Most of the changes are really cosmetics as "" -> <>, the actual changes have been discussed, and  everything seems fine with me

**Comment by kostrzewa** [comment]
Had to make one more modif, moving the non-generated file into the `include/` directory in the source dir. This is in order to be able to grab the settings when using tmLQCD as a library in an external program as in that case, one doesn't want to have to specify the source directory itself as an include dir (as there might be include files with the same names between the client app and tmLQCD). In the end, both `tmlqcd_config.h` and `tmlqcd_config_internal.h` should of course be installed.

I've checked this now linked against my cvc_depgraph extension of CVC and the changes merged in from master don't seem to break the QUDA behaviour.

**Comment by sunpho84** [comment]
> > what do you mean? isn't travis cehcking the build?
> 
> that has been broken for ages.

? travis is checking my fork and he seems to be rather happy...

https://travis-ci.com/github/sunpho84/tmLQCD/builds/222046903

**Pull request merged**